lib/xpm_ruby/client.rb

```diff
@@ -49,27 +49,9 @@ def build(**args)
     end
 
     def add(access_token:, xero_tenant_id:, client:)
-      response = Connection
+      Connection
         .new(access_token: access_token, xero_tenant_id: xero_tenant_id)
-        .post(endpoint: "client.api/add", data: client.to_xml(root: "Client"))
-
-      case response.status
-      when 401
-        hash = Ox.load(response.body, mode: :hash_no_attrs, symbolize_keys: false)
-
-        raise Unauthorized.new(hash["html"]["head"]["title"])
-      when 200
-        hash = Ox.load(response.body, mode: :hash_no_attrs, symbolize_keys: false)
-
-        case hash["Response"]["Status"]
-        when "OK"
-          hash["Response"]["Client"]
-        when "ERROR"
-          raise Error.new(response["ErrorDescription"])
-        end
-      else
-        raise Error.new(response.status)
-      end
+        .post(endpoint: "client.api/add", data: client.to_xml(root: "Client"))["Client"]
```

Comment:
Overall this is great (moving the error handling into the connection which is shared). Much needed! 👍 

Just a small thing, I am just wondering if it's a little more readable to still have a response variable

e.g.

```
response = Connection
        .new(access_token: access_token, xero_tenant_id: xero_tenant_id)
        .post(endpoint: "client.api/add", data: client.to_xml(root: "Client"))
response["Client"]
```

I am just wary that if we have too much method chaining the code can get a little hard to follow. We may even want a connection variable too

```
connection = Connection.new(access_token: access_token, xero_tenant_id: xero_tenant_id)
response = connection .post(endpoint: "client.api/add", data: client.to_xml(root: "Client"))
response["Client"]
```
But let me check the Ruby guidelines. This might just be a personal thing for me. ;)

Reply:
Like your first example Tony. Seems like a good compromise between too much chaining and too much assignment. I'll update.

Reply:
We can remove the chaining on `Connection` later by making it a module instead of an object.

Reply:
So I was doing some reading and it appears that method chaining is ok/good practice when you are working on the same object, but as these are returning different objects it can make debugging quite confusing.

e.g. if you were working with Arel where a query is always returned then `Person.where().order().etc()` where it is always working on the query object then it's good practice.

If each method returns a different object it can get confusing.

Elixir has a pipe operator (|>), but Elixir is functional which means you are only doing data transformations with each pipe which is slightly different.

We can have a chat about it in stand up though.
:)

Reply:
Yeap good points I got carried away - have updated :P